pkg/controllers/user/logging/clusterlogging.go

```diff
@@ -1,107 +1,118 @@
+		projectLoggingLister: cluster.Management.Management.ProjectLoggings("").Controller().Lister(),
 		roles:                cluster.RBAC.Roles(loggingconfig.LoggingNamespace),
 		rolebindings:         cluster.RBAC.RoleBindings(loggingconfig.LoggingNamespace),
-		clusterRoleBindings:  cluster.RBAC.ClusterRoleBindings(loggingconfig.LoggingNamespace),
-		clusterLister:        cluster.Management.Management.Clusters("").Controller().Lister(),
-		clusterName:          cluster.ClusterName,
+		services:             cluster.Core.Services(loggingconfig.LoggingNamespace),
+		serviceLister:        cluster.Core.Services("").Controller().Lister(),
+		serviceAccounts:      cluster.Core.ServiceAccounts(loggingconfig.LoggingNamespace),
 	}
-	clusterloggingClient.AddClusterScopedHandler("cluster-logging-controller", cluster.ClusterName, syncer.Sync)
+	clusterloggingClient.AddClusterScopedLifecycle("cluster-logging-controller", cluster.ClusterName, lifecycle)
 }
 
-func (c *ClusterLoggingSyncer) Sync(key string, obj *v3.ClusterLogging) error {
-	//clean up
-	if obj == nil || obj.DeletionTimestamp != nil {
-		return utils.CleanResource(c.namespaces, c.clusterLoggingLister, c.projectLoggingLister)
-	}
+func (c *ClusterLoggingLifecycle) Create(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
+	newObj, err := v3.LoggingConditionProvisioned.DoUntilTrue(obj, func() (runtime.Object, error) {
+		return obj, provision(c.namespaces, c.configmaps, c.serviceAccounts, c.clusterRoleBindings, c.daemonsets, c.clusterLister, c.clusterName)
+	})
+	return newObj.(*v3.ClusterLogging), err
+}
 
-	if err := utils.IniteNamespace(c.namespaces); err != nil {
-		return err
-	}
-	if err := utils.InitConfigMap(c.configmaps); err != nil {
-		return err
+func (c *ClusterLoggingLifecycle) Updated(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
+	newObj, err := v3.LoggingConditionProvisioned.DoUntilTrue(obj, func() (runtime.Object, error) {
+		return obj, provision(c.namespaces, c.configmaps, c.serviceAccounts, c.clusterRoleBindings, c.daemonsets, c.clusterLister, c.clusterName)
+	})
+	if err != nil {
+		return newObj.(*v3.ClusterLogging), err
 	}
 
-	//embedded
-	if utils.GetClusterTarget(obj.Spec) == "embedded" {
-		if err := utils.CreateOrUpdateEmbeddedTarget(c.deployments, c.serviceAccounts, c.services, c.roles, c.rolebindings, loggingconfig.LoggingNamespace, obj); err != nil {
-			return err
-		}
-		c.updateEmbeddedEndpoint()
-	} else {
-		if err := utils.RemoveEmbeddedTarget(c.deployments, c.serviceAccounts, c.services, c.roles, c.rolebindings); err != nil {
-			return err
-		}
+	cl := newObj.(*v3.ClusterLogging)
+	if reflect.DeepEqual(cl.Spec, obj.Status.AppliedSpec) {
+		return newObj.(*v3.ClusterLogging), nil
 	}
 
-	if err := c.createOrUpdateClusterConfigMap(); err != nil {
-		return err
+	newObj, err = v3.LoggingConditionUpdated.Do(newObj, func() (runtime.Object, error) {
+		return c.update(obj)
+	})
+
+	cl = newObj.(*v3.ClusterLogging)
+	if err != nil {
```

Comment:
I  think we do not have  to check err here, correct?

Reply:
It looks like it belongs to the line 96, before the casting happens.